File: src/main/java/com/github/lppedd/cc/CCBundle.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.github.lppedd.cc;

import java.lang.ref.Reference;
import java.lang.ref.SoftReference;
import java.util.ResourceBundle;

import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.PropertyKey;

import com.intellij.CommonBundle;

public final class CCBundle {
  @NonNls
  private static final String BUNDLE = "messages.ConventionalCommitBundle";
  private static Reference<ResourceBundle> bundleReference;

  public static String message(
      @NotNull @PropertyKey(resourceBundle = BUNDLE) final String key,
      @NotNull final Object... params) {
    return CommonBundle.message(getBundle(), key, params);
  }

  private static ResourceBundle getBundle() {
    // noinspection StaticVariableUsedBeforeInitialization
    ResourceBundle bundle = com.intellij.reference.SoftReference.dereference(bundleReference);

    if (bundle == null) {
      bundle = ResourceBundle.getBundle(BUNDLE);
      bundleReference = new SoftReference<>(bundle);
    }

    return bundle;
  }
}

File: src/main/kotlin/com/github/lppedd/cc/CCBundle.kt
package com.github.lppedd.cc

import com.intellij.BundleBase
import com.intellij.reference.SoftReference
import org.jetbrains.annotations.PropertyKey
import java.lang.ref.Reference
import java.util.*
import java.lang.ref.SoftReference as JavaSoftReference

/**
 * @author Edoardo Luppi
 */
public object CCBundle {
  private const val BUNDLE = "messages.ConventionalCommitBundle"

  private var bundleReference: Reference<ResourceBundle>? = null
  private val bundle: ResourceBundle
    get() = derefBundle()

  @JvmStatic
  public operator fun get(@PropertyKey(resourceBundle = BUNDLE) key: String, vararg params: Any): String =
    BundleBase.message(bundle, key, *params)

  @JvmStatic
  public fun getOrDefault(@PropertyKey(resourceBundle = BUNDLE) key: String, defaultValue: String, vararg params: Any): String =
    BundleBase.messageOrDefault(bundle, key, defaultValue, *params)

  private fun derefBundle(): ResourceBundle {
    var ref = SoftReference.dereference(bundleReference)

    if (ref == null) {
      ref = ResourceBundle.getBundle(BUNDLE)
      bundleReference = JavaSoftReference(ref)
    }

    return ref
  }
}
